core/src/main/java/org/apache/accumulo/core/singletons/SingletonManager.java

```diff
@@ -0,0 +1,189 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.accumulo.core.singletons;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+
+/**
+ * This class automates management of static singletons that maintain state for Accumulo clients.
+ * Historically, Accumulo client code that used Connector had no control over these singletons. The
+ * new AccumuloClient API that replaces Connector is closeable. When all AccumuloClients are closed
+ * then resources used by the singletons are released. This class coordinates releasing those
+ * resources. For compatibility purposes this class will not release resources when the user has
+ * created Connectors.
+ *
+ * <p>
+ * This class is intermediate solution to resource management. Ideally there would be no static
+ * state and AccumuloClients would own all of their state and clean it up on close. If
+ * AccumuloClient is not closable at inception, then it is harder to make it closable later. If
+ * AccumuloClient is not closable, then its hard to remove the static state. This class enables
+ * making AccumuloClient closable at inception so that static state can be removed later.
+ *
+ * @see org.apache.accumulo.core.util.CleanUp
+ */
+
+public class SingletonManager {
+
+  /**
+   * These enums determine the behavior of the SingletonManager.
+   *
+   */
+  public enum Mode {
+    /**
+     * In this mode singletons are disabled when the number of active client reservations goes to
+     * zero.
+     */
+    CLIENT,
+    /**
+     * In this mode singletons are never disabled.
+     */
+    SERVER,
+    /**
+     * In this mode singletons are never disabled unless the mode is set back to CLIENT. The user
+     * can do this by using util.CleanUp (an old API created for users).
+     */
+    CONNECTOR
+  }
+
+  private static long reservations;
+  private static Mode mode;
+  private static boolean enabled;
+  private static boolean transitionedFromClientToConnector;
+  private static List<SingletonService> services;
+
+  @VisibleForTesting
+  static void reset() {
+    reservations = 0;
+    mode = Mode.CLIENT;
+    enabled = true;
+    transitionedFromClientToConnector = false;
+    services = new ArrayList<>();
+  }
+
+  static {
+    // test code calls this, so always use it inorder to give credibility to test results
+    reset();
+  }
```

Comment:
Won't this get called when the class is loaded on the Server?  If this was for testing why have it in a static block?

Reply:
I removed the comment.  Was just trying to say that class initialization should use the same code as test.